[RFC] Deprecate `srb rbi` in favor of Tapioca

Morriar opened the pull request:
### Motivation

This PR proposes to deprecate the `srb rbi` command (and sub-commands) and encourages users to use [Tapioca](https://github.com/Shopify/tapioca) instead.

We think that Tapioca will provide a better experience than `srb rbi`:

* `srb` has caused [a lot of issues](https://github.com/sorbet/sorbet/labels/srb) in the past
* `srb` isn't used at Stripe and thus not made a priority
* Tapioca is now a mature solution used for many aspects of RBI generation at Shopify
* Tapioca already support most of the features from `srb rbi` (see table below)
* Tapioca provides [RBI generation for Rails and other DSL](https://github.com/Shopify/tapioca#generating-rbi-files-for-rails-and-other-dsls)

Here's a comparison of Tapioca and `srb rbi` commands:

| srb                          | tapioca                                           |
| :--------------------------: | :-----------------------------------------------: |
| `srb init`                   | [`tapioca init`][tapioca-init]                    |
| `srb rbi config`             | [`tapioca config`][tapioca-config]                |
| `srb rbi sorbet-typed`       | [`tapioca annotations`][tapioca-annotations]      |
| `srb rbi gems`               | [`tapioca gems`][tapioca-gems]                    |
| `srb rbi hidden-definitions` | Partially covered by [`tapioca dsl`][tapioca-dsl] |
| `srb rbi todo`               | [`tapioca todo`][tapioca-todo]                    |
| `srb rbi suggest-typed`      | Delegated to [`spoom bump`][spoom-bump]           |
| `srb rbi find-gem-rbis`      | Folded inside `tapioca gems`                      |

[tapioca-init]: https://github.com/Shopify/tapioca#getting-started
[tapioca-config]: https://github.com/Shopify/tapioca#getting-started
[tapioca-annotations]: https://github.com/Shopify/tapioca#pulling-rbi-annotations-from-remote-sources
[tapioca-gems]: https://github.com/Shopify/tapioca#generating-rbi-files-for-gems
[tapioca-dsl]: https://github.com/Shopify/tapioca#generating-rbi-files-for-rails-and-other-dsls
[tapioca-todo]: https://github.com/Shopify/tapioca#rbi-files-for-missing-constants-and-methods
[spoom-bump]: https://github.com/Shopify/spoom#change-the-sigil-used-in-files

### Changes

This PR only adds a few deprecation message and does not remove any feature.

In the wiki, we suggest to also install Tapioca when adopting Sorbet:

<img width="915" alt="image" src="https://user-images.githubusercontent.com/583144/181072978-9bdf9fcb-28ab-465d-ba30-a3f976f5991a.png">

Then we warn the user that the remaining documentation is about a deprecated feature:

<img width="915" alt="image" src="https://user-images.githubusercontent.com/583144/181073100-bfd02001-3aac-4a8b-9858-d6a00931c17e.png">

When running a `srb rbi` related command, the user now gets this warning:

~~~
$ bin/srb rbi sorbet-typed
⚠️  The `srb rbi` command is being deprecated.

This command will be removed in future versions and all RBI generation will be done through Tapioca.

Please use `tapioca` instead.

Add Tapioca to your Gemfile then run `bundle install` to install it:

```
gem "tapioca", require: false, :group => :development
```

Once Tapioca is installed, simply run `tapioca init` to initialize your project with Sorbet and generate the necessary RBI files:

```
$ bundle exec tapioca init
```

For more information, please see https://github.com/Shopify/tapioca.
~~~

What do you think?

jez commented on website/docs/adopting.md:
```diff
@@ -14,6 +14,7 @@ systems. We'll declare them in our Gemfile and install them with Bundler:
 
```
I can't comment on the above section of the diff, but:

```
There are two components to Sorbet: the command line interface and runtime
systems. We'll declare them in our Gemfile and install them with Bundler:
```

suggestion:

```
There are three components to Sorbet:

1. The static type checker, which runs as a command line executable (`srb tc`)
2. The runtime package, which provides syntax for type annotations as well as certain type-friendly data structures.
3. Tools for generating [RBI files](rbi.md), which allow Sorbet to interoperate with gems and metaprogramming.

We'll declare them in our Gemfile and install them with Bundler:
```

jez replied:
feel free to remix

jez commented on website/docs/adopting.md:
```diff
@@ -58,6 +60,30 @@ For small projects, Sorbet can run on a single file with no additional
 information. But for projects that have multiple files and depend on other gems,
 Sorbet needs to know more information to work.
 
+> **The `srb init` command is being deprecated.**
```
Do you mind if I draft a quick blog post which we could include in this PR, and could contain a longer discussion of what would change, why the change is being made, timelines, etc.?

paracycle reviewed (commented):
We probably want to change (or maybe completely remove) the post-install message from the `sorbet` gem that suggests folks run `srb init` to start working with Sorbet. It would be weird to see the command that the gem tells you to run telling you that it is deprecated.

jez commented:
I would also love if you switched from saying "`srb rbi` is deprecated" to either "tapioca is recommended over srb rbi"/"srb rbi is in maintenance mode" depending on the context.

I explained what maintenance mode will mean in the blog post, and you can feel free to link to that section header anywhere it would be convenient.

froydnj commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+For these reasons and many more, `srb rbi hidden-definitions` is one of the
+tools that people struggle with the most when adopting Sorbet.
+
+But its one selling point is that sometimes there's simply no other tool. It's
```
```suggestion
But its one selling point is that sometimes there's simply no other tool. Its
```

jez replied:
:shame:

froydnj commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+  effort to fix, we may or may not have time to invest in fixing it. For this
+  reason we recommend that you switch to Tapioca at your earliest convenience.
+
+Even with thousands of change to the Sorbet repo, the `srb rbi` command has only
```
```suggestion
Even with thousands of changes to the Sorbet repo, the `srb rbi` command has only
```

froydnj commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+  [custom DSL generators](https://github.com/Shopify/tapioca#writing-custom-dsl-compilers)
+  for teaching Sorbet about DSLs specific to your project.
+
+- Tapioca will do a better job at generating actually types inside the generated
```
```suggestion
- Tapioca will do a better job at generating actual types inside the generated
```

froydnj commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+## What's happening with `srb rbi hidden-definitions`?
+
+Tapioca has no exact replacement for `srb rbi hidden-definitions`. This is the
+main reason why we have no active plans to remove the `srb rbi` command and it's
```
```suggestion
main reason why we have no active plans to remove the `srb rbi` command and its
```

froydnj reviewed (commented):
Minor nits on the blog post, but otherwise seems reasonable!

Morriar commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+- The [`tapioca`](https://github.com/Shopify/tapioca) gem is now the recommended
+  approach to [RBI generation](/docs/rbi) in projects that use Sorbet. This
+  recommendation is reflected in the Sorbet docs. Tapioca replaces (and improves
+  on!) functionality currently provided by the `srb rbi` command. Tapioca is
```
Should it be `functionalities`?

```suggestion
  on!) functionalities currently provided by the `srb rbi` command. Tapioca is
```

jez replied:
I think both can be correct, I slightly prefer the way that "functionality" reads, as it presents the tool as a holistic approach, rather than a collection of small functionalities.

Morriar commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+if you need to reference it.
+
+If you'd like to start using Tapioca in a project that already uses `srb rbi`,
+[see this migration guide](#TODO).
```
You can point to https://github.com/Shopify/tapioca/wiki/Migrating-to-Tapioca

Morriar commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+[Tapioca](https://github.com/Shopify/tapioca) is a Ruby gem that provides
+command line tools for generating [RBI files](/docs/rbi). From its
+documentation, "Tapioca surfaces types and methods from many sources that Sorbet
+cannot otherwise see—such as gems, Rails and other DSLs—compiles them into RBI
```
```suggestion
cannot otherwise see — such as gems, Rails and other DSLs — compiles them into RBI
```

jez replied:
It doesn't render well, but these are em dashes:

<img width="806" alt="image" src="https://user-images.githubusercontent.com/5544532/181310498-38272e99-2b28-436e-8258-a86c36d56a3b.png">

which I prefer to use to typeset dashed, parenthetical expressions like this.

Morriar commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+alternative type checker for Ruby).
+
+Both Tapioca and `srb rbi` generate RBIs for a project's gems and support
+various generating RBI files that teach Sorbet about various kinds of
```
"support various generating RBI files" sounds weird. Should it be "support generating various RBI files"?

jez replied:
good catch thanks.

Morriar commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+
+1.  Write out an RBI containing that difference.
+
+This tool is great when it works—you don't have to think about it!—but extremely
```
```suggestion
This tool is great when it works — you don't have to think about it! — but extremely
```

jez replied:
ditto em dash comment above

Morriar commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -0,0 +1,266 @@
+  course, which will continue to be actively improved).
+
+- Minor changes that are easy to review will likely get reviewed (and
+  automatically released), but changes that merely add new features and are hard
```
```suggestion
  automatically released), but changes that merely add new features or are hard
```

Morriar reviewed (commented):
A few suggestions but this blogpost is amazing ❤️

Morriar commented on website/docs/rbi.md:
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
 ## The Hidden Definition RBI
 
+> **The `srb rbi hidden-definitions` is in
+> [maintenance mode](../blog/2022/07/28/srb-tapioca), please use `tapioca`
```
Is there a better way to link to blog posts?

jez replied:
I think this is likely best.

Morriar commented:
Update:

1. I removed the post install message
2. I updated the whole documentation to replace usages of `srb init` and friends by `tapioca`
3. I kept the `hidden-definitions` related section but moved the deprecation message there
4. I updated the deprecation message to use "maintenance mode" rather than "deprecated" and linked to the blog post
5. I added a section about `DSL RBIs`

There still are a few mentions of the `hidden-definitions` file but I don't want to remove them yet accordingly to @jez blogpost. 

I also didn't update the tests yet until we're all agreeing on the wording of the deprecation message.

jez commented:
@Morriar did you force push away the blog post? I don't see it anymore

jez commented on gems/sorbet/bin/srb-rbi:
```diff
@@ -37,6 +37,25 @@ module Sorbet::Private::Main
+    <<~MESSAGE
+      #{emojify("⚠️ ", yellow("The"))} #{cyan("srb rbi")} #{yellow("command is in maintenance mode, please use")} #{cyan("Tapioca")} #{yellow("instead.")}
+
+      #{yellow("The command")} #{cyan("srb rbi")} #{yellow("will be removed in future versions and all RBI generation should be done through Tapioca.")}
+
+      Add Tapioca to your Gemfile then run #{cyan("bundle install")} to install it:
```
```suggestion
      To switch, add Tapioca to your Gemfile then run #{cyan("bundle install")} to install it:
```

jez commented on website/docs/rbi.md:
```diff
@@ -28,19 +28,17 @@ We'll cover these questions in order.
-Note that `srb rbi update` is an alias for `srb init`.
+| I want to:                                                                 | so I'll run:                 |
+| -------------------------------------------------------------------------- | ---------------------------- |
+| Initialize a new Sorbet project, including all RBI files                   | `tapioca init`               |
+| &nbsp;                                                                     | &nbsp;                       |
+| Fetch pre-written RBIs from [`rbi-central`]                                | `tapioca rbi annotations`    |
+| (Re)generate RBIs for all gems using runtime reflection                    | `tapioca rbi gems`           |
+| (Re)generate RBIs for all DSLs such as Rails using runtime reflection      | `tapioca rbi dsl`            |
+| (Re)generate an RBI for all "hidden definitions" in a project (deprecated) | `srb rbi hidden-definitions` |
+| (Re)generate a the TODO RBI file (for missing constants)                   | `tapioca rbi todo`           |
```
Is `tapioca rbi` actually a command?

Morriar replied:
Good catch, thanks!

jez commented on website/docs/rbi.md:
```diff
@@ -126,14 +127,17 @@ The most common way to get RBI files for gems is to generate (or regenerate)
-# Update only the autogenerated gem RBIs:
-srb rbi gems
+# Update only the autogenerated gem RBIs for new gems:
+❯ bin/tapioca gems
```
What is `bin/tapioca`? Should this be `bundle exec tapioca`?

Morriar replied:
When you run `tapioca init`, Tapioca will create a [binstub](https://bundler.io/man/bundle-binstubs.1.html) for itself so you can run it through `bin/tapioca`

jez replied:
Great thanks, do you mind mentioning that `tapioca init` will create that in the `adopting.md` page?

Morriar replied:
Done in 4cf15c320 👍

Morriar commented on website/blog/2022-07-28-srb-tapioca.md:
```diff
@@ -116,6 +116,13 @@ Tapioca has some unique benefits compared to `srb rbi`:
+  [sorbet-typed](https://github.com/sorbet/sorbet-typed), which `srb rbi` uses.
+  The difference between the two repos is largely just about how metadata is
+  stored. It should be painless to contribute any relevant `sorbet-typed` RBI
+  files to `rbi-central` if your codebase depended on them.
```
There is also a difference on how we check the RBIs for correctness:

* Runtime checks, we ensure that every constant/method defined in the RBI actually exist at runtime
* Static checks: We ensure that 1) the definition is not redundant with what Tapioca produces (like adding an empty scope already generated by Tapioca, or a method without a signature) 2) doesn't conflict with what Tapioca generates (like changing the arity of a method) 3) that the generated RBI + the written one do not produce any error when running Sorbet on it.

jez replied:
I was trying to figure out how much to say "yeah it's just better" + link to the full docs, versus how much to explain in detail about why it's better.

Feel free to push whatever changes you'd like to this section.

Morriar replied:
Done in 36ae5d915 👍

jez commented on website/docs/rbi.md:
```diff
@@ -126,14 +127,17 @@ The most common way to get RBI files for gems is to generate (or regenerate)
+The way this works is that `tapioca` will load a project's Gemfile and `require`
 each gem declared within. While loading each gem, it uses runtime reflection to
 learn what things that gem defines, and then serializes this information into a
 single RBI file for that gem.
```
Can we add a note here to link to the tapioca docs about how to further customize the `require` process?

Morriar replied:
Done in 7243e6740 👍

jez commented on website/docs/rbi.md:
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
 
 Hand-written RBIs for gems can come from either:
 
-- [`sorbet-typed`], a central repository for sharing hand-written RBI files with
+- [`rbi-central`], a central repository for sharing hand-written RBI files with
```
Should we add a note here explaining rbi-central vs sorbet-typed?

Morriar replied:
Do you mean a deprecation warning or a link to this note in your blogpost?

jez replied:
I was thinking maybe just put something here like

```markdown
> **Note**: With [`srb rbi` moving to maintenance mode](blog-post), [`rbi-central`] has shifted to being the main repository of community-sourced RBI files for open source gems, replacing [`sorbet-typed`].
```

Morriar replied:
Done in f099f437d 👍

jez commented on website/docs/rbi.md:
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
 
 ## The Hidden Definition RBI
 
+> **The `srb rbi hidden-definitions` is in
```
```suggestion
> **The `srb rbi hidden-definitions` command is in
```

Morriar replied:
Done in 6f8a4ff4c 👍

Morriar commented on website/docs/rbi.md:
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
+
+Sorbet by itself does not understand DSLs involving meta-programming, such as
+Rails. This means that Sorbet won't know about constants and methods generated
+by `ActiveRecord` or `ActiveSupport`. To solve this, Tapioca can load your
```
Done in 49f6de5d9 👍
